### DISCUSSION TITLE
Cancellation not working as documented

### DISCUSSION THREAD

**Discussion by henda79** [Respawnsive/Apizr#14] [General]
Hey guys, first off this is great library... I'm glad I found it.

I've stated implementing this in .NET 7 MAUI and so far , so good, except I notice if I implement the cancellation like the docs, using the `optionsBuilder`, the call never gets cancelled.

https://www.apizr.net/articles/config_cancellation.html

However if I use `WithCancellation` on the `ExecuteAsync` method, it works as expected.

**Comment by JeremyBP** [comment]
Oops :s

Well, I have to admit that's quite the only feature I did not cover by tests, as you may notice into my Tests project. So classic...

Will try to find time to fix it this week, so expect a 5.3 shortly :)

Thanks for sharing

**Reply by henda79** [reply]
No problem :)

It's always the same and as I said, it works by attaching on the ExecuteAsync method which is more clean if not already using the optionsBuilder. Maybe mention that in the docs.

Thanks for the quick reply.

**Comment by JeremyBP** [comment]
Ok I just covered cancellation with some test cases and yes, Refit needs us to pass the CancellationToken into the request so that it could be cancelled on the Refit side and not only on the Apizr one.
So in order to cancel a request, we have to include explicitly the cancellation token into our request parameters like so:
![image](https://github.com/Respawnsive/Apizr/assets/4272307/d8d8178d-3c55-4ea9-ac36-31a1cfc34951)

Then, just pass it to the request:
![image](https://github.com/Respawnsive/Apizr/assets/4272307/21b6a6f7-fe06-4c07-913c-dc812d54367f)

> [!IMPORTANT]
> - The first Execute method will work and may be cancelled, but only on the Refit side as Apizr has no idea of the cancellation token in that case.
> - The second Execute method will work too and may be cancelled, but only on the Apizr side as Refit has no idea of the cancellation token in that case.
> - You'd better go with one of the ohters, providing the token to both Refit and Apizr. I would recommend  the opt.CancellationToken ones which guaranties you're passing the right token (it's not for now but it could be a linked one in the futur).

> [!NOTE] 
> Passing the options itself to the request is optional but recommended as it clarifies logs at least, even if you don't use the builder

I'll update the doc soon about CancellationToken.
Thanks for the focus on it.

**Comment by henda79** [comment]
No problem, glad I could help.

The code I was using is

```
await manager.ExecuteAsync(api => api.NetworkCall()).WithCancellation(ct.Token);
```

but I assume, now you've mentioned it, only cancels with Apizr and not Refit.

Are you saying I should add the cancellationToken on my Refit interface explicitly ? I used to have it like this but removed it after reading the docs :)

```
await manager.ExecuteAsync(
    api => api.NetworkCall(ct.Token),
    options => options.WithCancellation(ct.Token)
);
```

So something like that ?

**Reply by JeremyBP** [reply]
Yep, more or less back as usual actually. 
Better doing it like that:
```csharp
await manager.ExecuteAsync(
    (opt, api) => api.NetworkCall(opt.CancellationToken),
    options => options.WithCancellation(ct.Token)
);
```
It's all about Refit by design, which set a cancellation token at none if it couldn't find a token parameter from the method.
I'll dig into it, trying to find a way to provide a token to Refit without having to add token parameter into the interface method. If it's even possible.
Untill there, the updated doc is on its way.

**Reply by JeremyBP** [reply]
Wait! :)
It seems I finally managed to find something, soooo much quicker than I expected ^^
Actually, I (re)discovered that I've been asking myself about that limitation in the past (Apizr is a big fat boy now so I just can't remember it all since it was a baby).
Anyway, my code around cancellation management was half the way to the goal.
But now, after lunch (here in France and maybe a little glass of wine), I noticed what was stupidly missing.
Once fixed, my tests came green all the way by throwing a TaskCanceledException as expected and without having to pass the token as a Refit parameter \O/
So we're back to the future:
```csharp
await manager.ExecuteAsync(
    (opt, api) => api.NetworkCall(opt),
    options => options.WithCancellation(ct.Token)
);
```
I'm gonna push a 5.3-beta1 on Nuget so you could test it and tell me if it sounds good to you too.
I'll let you know when publish done, probably by the end of the week. 
I've got another boy, a 6y real one, who ask me to get out for now :)
Keep in touch

**Reply by henda79** [reply]
Sounds good... I wish I had a little glass of wine with lunch :) in the UK, it's more like a cup of tea at your desk, when working while it gets cold.

Yes family keeps me busy, I too have a young one, a bit older than yours though, but a girl so much higher demand :)

I'll keep a look out for it. I'm sure I'll keep in touch as I'll no doubt have questions as I integrate it.

**Reply by henda79** [reply]
Give me a nudge when this goes out, I noticed in testing an watching network activity, even though the cancellation happens in the code, refit still keeps retrying base don the policies.

Thanks.

**Comment by JeremyBP** [comment]
Hey @henda79,
Just pushed v5.3.0 on Nuget including the cancellation fix and the timeout new thing.
Could you please give it a try and then tell me backif it does fix your problem?

Thanks

**Reply by henda79** [reply]
Will do.
